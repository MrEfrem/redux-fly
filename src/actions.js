export const UPDATE_STATE = '@@COMPONENT_SET_STATE'
export const RESET_STATE = '@@COMPONENT_RESET_STATE'
