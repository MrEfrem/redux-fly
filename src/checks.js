import isPlainObject from 'lodash/isPlainObject'

/**
 * Check mountPath
 * @param  {string} path
 * @return {void}
 */
export const checkMountPath = (path) => {
  if (typeof path !== 'string' || !path.length) {
    throw new Error('Mount path must be string')
  }
}

/**
 * Check initialState
 * @param  {Object} initialState
 * @return {void}
 */
export const checkInitialState = (initialState) => {
  if (!isPlainObject(initialState)) {
    throw new Error('InitialState must be plain object')
  }
}

/**
 * Check listenActions
 * @param  {Object} listenActions
 * @return {void}
 */
export const checkListenActions = (listenActions) => {
  if (listenActions && !isPlainObject(listenActions)) {
    throw new Error('ListenActions must be plain object')
  }
}

/**
 * Check options
 * @param  {Object} options
 * @return {void}
 */
export const checkOptions = (options) => {
  if (!isPlainObject(options)) {
    throw new Error('Options must be plain object ')
  }
}

/**
 * Check properties of an object options
 * @param  {Object}  options
 * @return {void}
 */
export const checkDetailOptions = (options) => {
  if (typeof options.connectToStore !== 'boolean') {
    throw new Error('connectToStore property of an object options must be boolean')
  }
  if (typeof options.unregisterInUnmount !== 'boolean') {
    throw new Error('Property unregisterInUnmount in third parameter shall be boolean')
  }
}
