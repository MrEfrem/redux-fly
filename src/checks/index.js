import isPlainObject from 'lodash/isPlainObject'

/**
 * Check mountPath
 * @param  {string} path
 * @return {void}
 */
export const checkMountPath = (path) => {
  if (typeof path !== 'string' || !path.length) {
    throw new Error('Mount path must be string')
  }
}

/**
 * Check initialState
 * @param  {Object} initialState
 * @return {void}
 */
export const checkInitialState = (initialState) => {
  if (!isPlainObject(initialState)) {
    throw new Error('InitialState must be plain object')
  }
}

/**
 * Check options
 * @param  {Object} options
 * @return {void}
 */
export const checkOptions = (options) => {
  if (!isPlainObject(options)) {
    throw new Error('Options must be plain object ')
  }
}
