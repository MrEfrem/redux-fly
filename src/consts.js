// @flow
export const RESET_STATE = '@@COMPONENT_RESET_STATE'
export const PROCESS_BATCH = '@@COMPONENT_PROCESS_BATCH'
export const MOUNT_PATH = '@@MOUNT_PATH'
export const UUID = '@@UUID'
export const BATCH = '@@BATCH'
export const NEW_STATE = '@@NEW_STATE'
export const ACTIONS = '@@ACTIONS'
export const PROP_MOUNT_PATH = 'reduxMountPath'
export const PROP_PERSIST = 'reduxPersist'
