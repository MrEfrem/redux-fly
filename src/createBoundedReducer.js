// @flow
import { RESET_STATE, MOUNT_PATH, NEW_STATE } from './consts'

/**
 * Create reducer bounded on mountPath
 * @param {string} mountPath
 * @param {Object} initialState
 * @param {Object} listenActions
 * @param {string} actionPrefix
 * @return {Function}
 *   @param {Object = initialState} state
 *   @param {Object} action
 *   @return {Object} new state
 */
export default (mountPath: string, initialState: Object, listenActions: Object, actionPrefix: string) =>
  (state: Object = initialState, action: Object) => {
    const resetState = `${actionPrefix}/${RESET_STATE}`
    if ((typeof action[MOUNT_PATH] !== 'undefined' && action[MOUNT_PATH] === mountPath && typeof action[NEW_STATE] !== 'undefined') ||
      action.type in listenActions
    ) {
      // Action is reset state
      if (action.type === resetState) {
        return { ...action[NEW_STATE] }
      }
      // Action found in listenActions
      const reducer = listenActions[action.type]
      if (reducer) {
        return {
          ...state,
          ...reducer(state, action)
        }
      }
      // Action dispatched by setReduxState
      return {
        ...state,
        ...action[NEW_STATE]
      }
    }
    return state
  }
