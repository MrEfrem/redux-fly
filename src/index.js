// @flow
export { default as createStore } from './createStore'
export { default as registerReducers } from './registerReducers'
export { default as createReducer } from './createReducer'
export { default as getState } from './getState'

import { MOUNT_PATH } from './consts'
export const consts = { MOUNT_PATH }
