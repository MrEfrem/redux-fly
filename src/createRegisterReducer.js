// @flow
import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import storeShape from './utils/storeShape'
import createBoundedReducer from './createBoundedReducer'
import setReduxState from './setReduxState'
import { RESET_STATE, MOUNT_PATH, UUID } from './consts'
import getState from './getState'
import genUUIDv4 from './genUUIDv4'

/**
 * Create and mount reducer
 * @param  {string} mountPath
 * @param  {Object} preloadedState
 * @param  {Object} listenActions
 * @param  {Object} options
 * @param {Object} wrapped React component
 * @return {Object} React component
 */
export default function createRegisterReducer(mountPath: string, preloadedState: Object, listenActions?: Object, options: Object,
  WrappedComponent: any
) {
  const { connectToStore, persist, actionPrefix } = options
  let ChildComponent = WrappedComponent
  if (connectToStore) {
    ChildComponent = connect((state) =>
      ({ reduxState: getState(mountPath)(state) })
    )(WrappedComponent)
  }
  const uuid = genUUIDv4()
  return class RegisterReducer extends React.Component {
    static contextTypes = {
      store: process.env.NODE_ENV === 'test' ? PropTypes.object : storeShape
    }

    reducer: ?Object
    setReduxState: any
    resetReduxState: any
    batchActions: any

    componentWillMount() {
      const { store } = this.context
      if (typeof store !== 'object') {
        throw new Error('Redux store must be created')
      }
      if (typeof store.registerReducers !== 'function') {
        throw new Error('Redux store must be enhanced with redux-fly')
      }
      this.reducer = {
        [mountPath]: createBoundedReducer(uuid, mountPath, preloadedState, listenActions || {}),
      }
      // Registration of created reducer
      store.registerReducers(this.reducer)
      // Binding setReduxState with redux store
      this.setReduxState = setReduxState(uuid, mountPath, store.dispatch, store.getState, actionPrefix)
      // Action creator RESET redux state
      this.resetReduxState = () => store.dispatch({
        type: RESET_STATE,
        [MOUNT_PATH]: mountPath,
        [UUID]: uuid
      })
    }

    componentWillUnmount() {
      // If component isn't persist then RESET redux state
      if (!persist) {
        this.resetReduxState()
      }
      this.reducer = null
      this.setReduxState = null
      this.resetReduxState = null
      this.batchActions = null
    }

    render() {
      return (
        <ChildComponent
          {...this.props}
          setReduxState={this.setReduxState}
          resetReduxState={this.resetReduxState}
          reduxMountedPath={mountPath}
          persist={process.env.NODE_ENV === 'test' ? persist : null}
        />
      )
    }
  }
}
