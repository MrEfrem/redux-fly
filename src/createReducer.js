// @flow
import React, { PropTypes } from 'react'
import { checkMountPath, checkPreloadedState, checkOptions } from './utils/checks'
import createRegisterReducer from './createRegisterReducer'
import isPlainObject from 'lodash/isPlainObject'
import warning from './utils/warning'
import { PROP_MOUNT_PATH, PROP_PERSIST } from './consts'
import { normalizeMountPath } from './utils/normalize'

/**
 * Check listenActions
 */
const checkListenActions = (listenActions) => {
  if (!isPlainObject(listenActions)) {
    throw new Error('ListenActions must be plain object')
  }
}

/**
 * Check persist
 */
const checkPersist = (persist) => {
  if (typeof persist !== 'boolean') {
    throw new Error('Persist must be boolean')
  }
}

/**
 * Check properties of an object options
 * @param  {Array}  default options
 * @param  {Object}  options
 * @return {void}
 */
const checkDetailOptions = (defaultOptions, options) => {
  if (typeof options.connectToStore !== 'boolean') {
    throw new Error('ConnectToStore must be boolean')
  }
  checkPersist(options.persist)
  if (typeof options.actionPrefix !== 'string') {
    throw new Error('ActionPrefix must be string')
  }
  if (['production', 'test'].indexOf(process.env.NODE_ENV) === -1) {
    const undefinedOptions = Object.keys(options).reduce((prev, next) => {
      if (defaultOptions.indexOf(next) === -1) {
        prev = `${prev}${next}, `
      }
      return prev
    }, '').slice(0, -2)
    if (undefinedOptions) {
      warning(`Undefined options: ${undefinedOptions}`)
    }
  }
}

/**
 * Create/mount reducer
 * @param  {string} mountPath
 * @param  {Object} preloadedState
 * @param  {Object} listenActions
 * @param  {Object} options
 * @return {
 *   @param {Object} wrapped React component
 *   @return {Object} React component
 * }
 */
export default (
  mountPath?: string,
  preloadedState: Function | Object,
  listenActions?: Function | Object,
  options: Object = {}
) => {
  if (mountPath || mountPath !== null) {
    checkMountPath(mountPath)
  }

  if (typeof preloadedState !== 'function') {
    checkPreloadedState(preloadedState)
  }

  if (listenActions && typeof listenActions !== 'function') {
    checkListenActions(listenActions)
  }

  checkOptions(options)

  const defaultOptions = {
    connectToStore: true,
    persist: false,
    actionPrefix: '',
  }
  const _options = {
    ...defaultOptions,
    ...options
  }
  checkDetailOptions(Object.keys(defaultOptions), _options)

  return (WrappedComponent: any) =>
    // Transferred some parameters is functions
    class CreateReducer extends React.Component {
      RegisterReducer: any
      propMountPath: ?string

      static propTypes = {
        [PROP_MOUNT_PATH]: process.env.NODE_ENV === 'test' ? PropTypes.any : PropTypes.string,
        [PROP_PERSIST]: process.env.NODE_ENV === 'test' ? PropTypes.any : PropTypes.bool
      }

      componentWillMount() {
        let { [PROP_MOUNT_PATH]: propMountPath, [PROP_PERSIST]: propPersist } = this.props
        // Mount path must be passed in props or options
        if (!mountPath && !propMountPath) {
          throw new Error('Mount path must be defined')
        }

        let realMountPath = mountPath
        // Priority mount path from props
        if (typeof propMountPath !== 'undefined') {
          checkMountPath(propMountPath)
          realMountPath = propMountPath
        }
        realMountPath = normalizeMountPath(realMountPath)

        // Priority persist from props
        if (typeof propPersist !== 'undefined') {
          checkPersist(propPersist)
          _options.persist = propPersist
        }

        // Preloaded state is function
        let _preloadedState = preloadedState
        if (typeof _preloadedState === 'function') {
          _preloadedState = _preloadedState(this.props, realMountPath)
          checkPreloadedState(_preloadedState)
        }

        // Listen actions is function
        let _listenActions = listenActions
        if (typeof _listenActions === 'function') {
          _listenActions = _listenActions(this.props, realMountPath)
          checkListenActions(_listenActions)
        }

        this.RegisterReducer = createRegisterReducer(realMountPath, _preloadedState, _listenActions,
          _options, WrappedComponent)
      }

      componentWillUnmount() {
        this.RegisterReducer = null
      }

      render() {
        const RegisterReducer = this.RegisterReducer
        return <RegisterReducer {...this.props} />
      }
    }
}
