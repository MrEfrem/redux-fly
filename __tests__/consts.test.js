import * as consts from '../src/consts'

test('Verify Defined CONSTS', () => {
  expect(consts).toMatchSnapshot()  
})
