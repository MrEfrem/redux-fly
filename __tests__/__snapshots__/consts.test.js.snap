exports[`test Verify Defined CONSTS 1`] = `
Object {
  "ACTIONS": "@@ACTIONS",
  "BATCH": "@@BATCH",
  "MOUNT_PATH": "@@MOUNT_PATH",
  "NEW_STATE": "@@NEW_STATE",
  "PROCESS_BATCH": "@@COMPONENT_PROCESS_BATCH",
  "PROP_MOUNT_PATH": "reduxMountPath",
  "PROP_PERSIST": "reduxPersist",
  "RESET_STATE": "@@COMPONENT_RESET_STATE",
  "UUID": "@@UUID",
}
`;
